### ISSUE TITLE
Use 2048-bit custom Diffie-Hellman primes

### ISSUE THREAD

**Issue by yunohost-bot** [YunoHost/issues#93]
###### Original Redmine Issue: [92](https://dev.yunohost.org/issues/92)

Author Name: **Bouska**

---

Currently OpenSSL uses 1024-bit non-custom Diffie-Hellman primes. This is considered as unsecure, see https://weakdh.org/

weakdh.org provides a [documentation](https://weakdh.org/sysadmin.html) to fix this security issue. By default, this should concern (at least) OpenSSH, Postfix, Dovecot, Metronome, and nginx.

Blocks issue #91

**Comment by frju365** [comment]
Was it on purpose to close this issue after PR #949 ? :-) I don't think this issue is fixed yet.

(it reminds me I had a PR in waiting for test I was suppose to do, so let's go tomorrow)

**Comment by alexAubin** [comment]
Ugh no indeed, not on purpose, thanks for the head up !

**Comment by SohKa** [comment]
Based on [RFC 7919](https://tools.ietf.org/html/rfc7919), [Mozilla recommends](https://github.com/mozilla/ssl-config-generator/issues/60) to use predefined DH groups and not custom ones (the ones generated with the `openssl` command). [In Mozilla's wiki](https://wiki.mozilla.org/Security/Archive/Server_Side_TLS_4.0#Pre-defined_DHE_groups), they reaffirm the advice. [The tool ssl-config.mozilla.org](https://ssl-config.mozilla.org/#server=nginx&version=1.10.3&config=intermediate&openssl=1.1.0l&guideline=5.4) gives this instruction for an nginx intermediate config (the one that Yunohost uses) :
```bash
    # curl https://ssl-config.mozilla.org/ffdhe2048.txt > /path/to/dhparam
    ssl_dhparam /path/to/dhparam;
```
In our situation, we can do this (as root) : 
```bash
curl https://ssl-config.mozilla.org/ffdhe2048.txt > /etc/ssl/private/dhparams.pem
```
Or [from this git repo](https://github.com/internetstandards/dhe_groups)
```bash
curl -L https://github.com/internetstandards/dhe_groups/raw/master/ffdhe2048.pem > /etc/ssl/private/dhparams.pem
```
And add this parameter to nginx's conf (in `/etc/nginx/conf.d/security.conf.inc`) : 
```
ssl_dhparam /etc/ssl/private/dhparams.pem
```
Because the `ffdhe2048` group will progressively phase out, I advise to directly use `ffdhe3072`.

I would love to do a PR to implement this but my Python sucks... But if nobody is available, I can take the time to learn. :)

**Comment by alexAubin** [comment]
>Based on RFC 7919, Mozilla recommends to use predefined DH groups and not custom ones (the ones generated with the openssl command). In Mozilla's wiki, they reaffirm the advice.

Wow I was not expecting this o.O But that's "good" to hear as it simplifies the logistic of having to generate a custom DH param.

I'm not sure exactly how to proceed, maybe we could ship this file directly in Yunohost's code (we already do this for other things like list of most used password)

~~Or maybe there's a debian repo that ships this file somehow ?~~ nvm the file is pretty small in fact

**Comment by alexAubin** [comment]
@SohKa  (or anybody else) : so if you're motivated, I don't think you need to know any python to do this. Naively I would : 
- add the .pem file [here](https://github.com/YunoHost/yunohost/tree/stretch-unstable/data/other)
- decide where this .pem file should be put after yunohost's installation (c.f. e.g. [here](https://github.com/YunoHost/yunohost/blob/stretch-unstable/debian/install#L10) for case) - maybe somewhere in /etc/ssl like Mozilla suggests
- (on dev env, you will need to manually cp the file where you want it to be unless you really want to compile/reinstall a proper .deb but that's a small technical detail)
- enable the line [here](https://github.com/YunoHost/yunohost/blob/stretch-unstable/data/templates/nginx/security.conf.inc#L19)
- test it by running `yunohost tools regen-conf nginx`

**Comment by SohKa** [comment]
@alexAubin Ok, will do !

**Comment by SohKa** [comment]
The fix has been released in the version 3.8.4.8 of yunohost. I just did the upgrade and it works fine. :)
I think this ticket can now be closed.
